File: api/persona_id/route.py
# api/persona_id/route.py — GET/PATCH/DELETE /api/persona/{id}
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from .._shared.db import get_cursor
from .._shared.auth import get_current_user
import json, re

router = APIRouter()

def _extract_id(path: str) -> str:
    m = re.search(r"/persona/([\w-]+)", path)
    return m.group(1) if m else ""

@router.get("/persona/{persona_id}")
def get_persona(persona_id: str, request: Request):
    try:
        auth = request.headers.get("authorization", "")
        user = get_current_user(auth)
    except ValueError as e:
        return JSONResponse({"error": str(e)}, status_code=401)

    with get_cursor() as c:
        c.execute("SELECT * FROM personas WHERE persona_id=? AND user_id=?", (persona_id, user["user_id"]))
        row = c.fetchone()

    if not row:
        return JSONResponse({"error": "不存在"}, status_code=404)

    result = dict(row)
    result["extracted_persona"] = json.loads(result.get("extracted_persona", "{}"))
    result["chat_data"] = json.loads(result.get("chat_data", "{}"))
    return result

@router.patch("/persona/{persona_id}")
def patch_persona(persona_id: str, request: Request):
    try:
        auth = request.headers.get("authorization", "")
        user = get_current_user(auth)
    except ValueError as e:
        return JSONResponse({"error": str(e)}, status_code=401)

    body = json.loads(request.body.decode())
    fields, values = [], []
    for field in ("name", "description", "extracted_persona"):
        if field in body:
            fields.append(f"{field}=?")
            val = body[field]
            if field == "extracted_persona":
                val = json.dumps(val, ensure_ascii=False)
            values.append(val)

    if not fields:
        return JSONResponse({"error": "没有要更新的字段"}, status_code=400)

    values.extend([persona_id, user["user_id"]])
    with get_cursor() as c:
        c.execute(f"UPDATE personas SET {', '.join(fields)} WHERE persona_id=? AND user_id=?", values)
        if c.rowcount == 0:
            return JSONResponse({"error": "不存在"}, status_code=404)

    return {"success": True}

@router.delete("/persona/{persona_id}")
def delete_persona(persona_id: str, request: Request):
    try:
        auth = request.headers.get("authorization", "")
        user = get_current_user(auth)
    except ValueError as e:
        return JSONResponse({"error": str(e)}, status_code=401)

    with get_cursor() as c:
        c.execute("DELETE FROM personas WHERE persona_id=? AND user_id=?", (persona_id, user["user_id"]))
        if c.rowcount == 0:
            return JSONResponse({"error": "不存在"}, status_code=404)
        c.execute("DELETE FROM chat_history WHERE persona_id=?", (persona_id,))

    return {"success": True}

File: frontend/app/api/persona/[id]/route.ts
// app/api/persona/[id]/route.ts — GET/PATCH/DELETE /api/persona/{id}
import { NextRequest, NextResponse } from 'next/server'
import { getDb } from '../../../_shared/db'
import { getCurrentUser } from '../../../_shared/auth'

export async function GET(
  request: NextRequest,
  { params }: { params: Promise<{ id: string }> }
) {
  try {
    const { id } = await params
    const user = getCurrentUser(request.headers.get('authorization'))
    const db = getDb()
    const row = db.prepare(
      'SELECT * FROM personas WHERE persona_id=? AND user_id=?'
    ).get(id, user.user_id) as any
    if (!row) return NextResponse.json({ error: '不存在' }, { status: 404 })
    return NextResponse.json({
      ...row,
      extracted_persona: JSON.parse(row.extracted_persona || '{}'),
      chat_data: JSON.parse(row.chat_data || '{}')
    })
  } catch (e: any) {
    const status = e.message?.includes('Token') ? 401 : 500
    return NextResponse.json({ error: e.message }, { status })
  }
}

export async function PATCH(
  request: NextRequest,
  { params }: { params: Promise<{ id: string }> }
) {
  try {
    const { id } = await params
    const user = getCurrentUser(request.headers.get('authorization'))
    const body = await request.json()
    const db = getDb()
    const fields: string[] = []
    const vs: any[] = []
    for (const [k, v] of Object.entries(body)) {
      if (['name', 'description', 'extracted_persona'].includes(k)) {
        fields.push(`${k}=?`)
        vs.push(k === 'extracted_persona' ? JSON.stringify(v) : v)
      }
    }
    if (!fields.length) return NextResponse.json({ error: '没有要更新的字段' }, { status: 400 })
    vs.push(id, user.user_id)
    const result = db.prepare(
      `UPDATE personas SET ${fields.join(',')} WHERE persona_id=? AND user_id=?`
    ).run(...vs)
    if (result.changes === 0) return NextResponse.json({ error: '不存在' }, { status: 404 })
    return NextResponse.json({ success: true })
  } catch (e: any) {
    return NextResponse.json({ error: e.message }, { status: 500 })
  }
}

export async function DELETE(
  request: NextRequest,
  { params }: { params: Promise<{ id: string }> }
) {
  try {
    const { id } = await params
    const user = getCurrentUser(request.headers.get('authorization'))
    const db = getDb()
    const result = db.prepare(
      'DELETE FROM personas WHERE persona_id=? AND user_id=?'
    ).run(id, user.user_id)
    if (result.changes === 0) return NextResponse.json({ error: '不存在' }, { status: 404 })
    db.prepare('DELETE FROM chat_history WHERE persona_id=?').run(id)
    return NextResponse.json({ success: true })
  } catch (e: any) {
    return NextResponse.json({ error: e.message }, { status: 500 })
  }
}
